### PR TITLE
Updated hf_argparser.py

### LINKED ISSUE #19116: HfArgumentParser support yaml parser
### Feature request

HfArgumentParser now supports for parsing dict and json files, will it be possible to support for parsing the widely used yaml files?

### Motivation

I think using yaml is a good way to record arguments.

### Your contribution

Not yet.

### PR DESCRIPTION
**Pull request by IMvision12**
# What does this PR do?

Works on #19116

## Summary
- [x] adding as parse_yaml_file method to HfArgumentParser with the code above
- [x] refactor the dupe code between parse_json_file and parse_dict similar to the code above
- [ ] add a small test of parse_yaml_file
- [ ] add a small test of parse_json_file

### CODE REVIEW

**Comment by sgugger** [comment]
Thanks for your PR! Can you just run `make style` on your branch to fix the code quality issue?

**Comment by IMvision12** [comment]
Yeah Sure!

File: src/transformers/hf_argparser.py [hunk 1]
```diff
@@ -234,60 +236,70 @@ def parse_args_into_dataclasses(
...
-        outputs = []
-        for dtype in self.dataclass_types:
-            keys = {f.name for f in dataclasses.fields(dtype) if f.init}
-            inputs = {k: v for k, v in args.items() if k in keys}
-            unused_keys.difference_update(inputs.keys())
-            obj = dtype(**inputs)
-            outputs.append(obj)
-        if not allow_extra_keys and unused_keys:
-            raise ValueError(f"Some keys are not used by the HfArgumentParser: {sorted(unused_keys)}")
+        outputs = self.parse_dict(json.loads(Path(json_file).read_text()))
```
Comment (sgugger) [inline src/transformers/hf_argparser.py:284]:
```suggestion
        outputs = self.parse_dict(json.loads(Path(json_file).read_text()), allow_extra_keys=allow_extra_keys)
```

File: src/transformers/hf_argparser.py [hunk 2]
```diff
@@ -234,60 +236,70 @@ def parse_args_into_dataclasses(
...
+            allow_extra_keys (`bool`, *optional*, defaults to `False`):
+                Defaults to False. If False, will raise an exception if the json file contains keys that are not
+                parsed.
+
+        Returns:
+            Tuple consisting of:
+
+                - the dataclass instances in the same order as they were passed to the initializer.
+        """
+        outputs = self.parse_dict(yaml.safe_load(yaml_file))
```
Comment (sgugger) [inline src/transformers/hf_argparser.py:304]:
```suggestion
        outputs = self.parse_dict(yaml.safe_load(yaml_file), allow_extra_keys=allow_extra_keys)
```

**Review by sgugger — APPROVED** [review]
Looking good, just make sure to pass along the `allow_extra_keys` argument :-)

**Comment by IMvision12** [comment]
Done @sgugger!!

**Comment by sgugger** [comment]
Thanks again!

**Pull request merged**